module/data/advancement/ability-score-improvement.mjs

```diff
@@ -0,0 +1,50 @@
+import { SparseDataModel } from "../abstract.mjs";
+import { MappingField } from "../fields.mjs";
+
+/**
+ * Data model for the Ability Score Improvement advancement configuration.
+ *
+ * @property {number} points                 Number of points that can be assigned to any score.
+ * @property {Object<string, number>} fixed  Number of points automatically assigned to a certain score.
+ */
+export class AbilityScoreImprovementConfigurationData extends foundry.abstract.DataModel {
+  /** @inheritdoc */
+  static defineSchema() {
+    return {
+      // TODO: This should default to 2 if added to a class, or 0 if added to anything else
+      points: new foundry.data.fields.NumberField({
+        integer: true, min: 0, initial: 2,
+        label: "DND5E.AdvancementAbilityScoreImprovementPoints",
+        hint: "DND5E.AdvancementAbilityScoreImprovementPointsHint"
+      }),
+      fixed: new MappingField(
+        new foundry.data.fields.NumberField({nullable: false, integer: true, initial: 0}),
+        {label: "DND5E.AdvancementAbilityScoreImprovementFixed"}
+      )
+    };
+  }
+}
+
+/**
+ * Data model for the Ability Score Improvement advancement value.
+ *
+ * @property {string} type  When on a class, whether the player chose ASI or a Feat.
+ * @property {Object<string, number>}  Points assigned to individual scores.
+ * @property {Object<string, string>}  Feat that was selected.
+ */
+export class AbilityScoreImprovementValueData extends SparseDataModel {
+  /** @inheritdoc */
+  static defineSchema() {
+    return {
+      type: new foundry.data.fields.StringField({
+        required: true, initial: "asi", choices: ["asi", "feat"]
+      }),
+      assignments: new MappingField(new foundry.data.fields.NumberField({
+        nullable: false, integer: true
+      }), {required: false, initial: undefined}),
+      feat: new MappingField(new foundry.data.fields.StringField(), {
```

Comment:
The fact that this is a `MappingField` is awkward, but I understand that it fits better into the rest of our framework around adding items via advancements.